nsxt/resource_nsxt_failure_domain.go

```diff
@@ -0,0 +1,156 @@
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx/model"
+)
+
+var edgeServices = []string{
+	"active",
+	"standby",
+	"no_preference",
+}
+
+func resourceNsxtFailureDomain() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceNsxtFailureDomainCreate,
+		Read:   resourceNsxtFailureDomainRead,
+		Update: resourceNsxtFailureDomainUpdate,
+		Delete: resourceNsxtFailureDomainDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"display_name": getDisplayNameSchema(),
+			"description":  getDescriptionSchema(),
+			"revision":     getRevisionSchema(),
+			"tag":          getTagsSchema(),
+			"preferred_edge_services": {
+				Type:         schema.TypeString,
+				Description:  "Set preference for failure domain",
+				Optional:     true,
+				Default:      "no_preference",
+				ValidateFunc: validation.StringInSlice(edgeServices, false),
+			},
+		},
+	}
+}
+
+func resourceNsxtFailureDomainRead(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+
+	id := d.Id()
+	if id == "" {
+		return fmt.Errorf("Error obtaining FailureDomain ID")
+	}
+
+	client := nsx.NewFailureDomainsClient(connector)
+	obj, err := client.Get(id)
+	if err != nil {
+		return fmt.Errorf("error during Failure Domain read: %v", err)
+	}
+
+	d.Set("display_name", obj.DisplayName)
+	d.Set("description", obj.Description)
+	setMPTagsInSchema(d, obj.Tags)
+	d.Set("revision", obj.Revision)
+
+	preferPtr := obj.PreferredActiveEdgeServices
+	preferStr := "no_preference"
+	if preferPtr != nil {
+		if *preferPtr == true {
+			preferStr = "active"
+		} else {
+			preferStr = "standby"
+		}
+	}
+	d.Set("preferred_edge_services", preferStr)
+	return nil
+}
+
+func failureDomainSchemaToModel(d *schema.ResourceData) model.FailureDomain {
+	displayName := d.Get("display_name").(string)
+	description := d.Get("description").(string)
+	tags := getMPTagsFromSchema(d)
+
+	obj := model.FailureDomain{
+		DisplayName: &displayName,
+		Description: &description,
+		Tags:        tags,
+	}
+
+	preferredEdgeServices := d.Get("preferred_edge_services").(string)
+	if preferredEdgeServices != "no_preference" {
+		activePrefer := true
+		standbyPrefer := false
+		if preferredEdgeServices == "active" {
+			obj.PreferredActiveEdgeServices = &activePrefer
+		} else {
+			obj.PreferredActiveEdgeServices = &standbyPrefer
+		}
+	}
+	return obj
+}
+
+func resourceNsxtFailureDomainCreate(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := nsx.NewFailureDomainsClient(connector)
+
+	failureDomain := failureDomainSchemaToModel(d)
+	obj, err := client.Create(failureDomain)
+	if err != nil {
+		id := ""
+		if obj.Id != nil {
+			id = *obj.Id
+		}
+		return handleCreateError("Failure Domain", id, err)
+	}
+	log.Printf("[INFO] Creating Failure Domain with ID %s", *obj.Id)
```

Comment:
This log should probably move to location before API call, for consistency

Reply:
The reason why I put it here is the ID can only be known after the API call, like the edge cluster resource. Let me know if you think moving it prior to the API call with `description_name` as a parameter is a better way.

Reply:
Makes sense, lets rename `Creating` to `Created` then

Reply:
Done.